hacspec/src/lib.rs

```diff
@@ -855,46 +884,62 @@ fn parse_message_1(
     EDHOCError,
 > {
     let mut error: EDHOCError = EDHOCError::UnknownError;
+    let mut method: U8 = U8(0xff);
     let mut g_x: BytesP256ElemLen = BytesP256ElemLen::new();
     let mut suites_i = BytesSuites::new();
     let mut suites_i_len: usize = 0;
     let mut raw_suites_len: usize = 0;
     let mut c_i = U8(0);
     let mut ead_1 = None::<EADItemHacspec>;
 
-    let method = rcvd_message_1.content[0];
-
-    let res_suites = parse_suites_i(rcvd_message_1);
+    // check surplus array encoding
+    if !is_cbor_array(rcvd_message_1.content[0]) {
+        method = rcvd_message_1.content[0];
 
-    if res_suites.is_ok() {
-        (suites_i, suites_i_len, raw_suites_len) = res_suites.unwrap();
+        let res_suites = parse_suites_i(rcvd_message_1);
 
-        g_x = BytesP256ElemLen::from_slice(
-            &rcvd_message_1.content,
-            3 + raw_suites_len,
-            P256_ELEM_LEN,
-        );
+        if res_suites.is_ok() {
+            (suites_i, suites_i_len, raw_suites_len) = res_suites.unwrap();
 
-        c_i = rcvd_message_1.content[3 + raw_suites_len + P256_ELEM_LEN];
+            let g_x_type = rcvd_message_1.content[1 + raw_suites_len].declassify();
+            if g_x_type == CBOR_BYTE_STRING {
+                g_x = BytesP256ElemLen::from_slice(
+                    &rcvd_message_1.content,
+                    3 + raw_suites_len,
+                    P256_ELEM_LEN,
+                );
 
-        // if there is still more to parse, the rest will be the EAD_1
-        if rcvd_message_1.len > (4 + raw_suites_len + P256_ELEM_LEN) {
-            // NOTE: since the current implementation only supports one EAD handler,
-            // we assume only one EAD item
-            let ead_res = parse_ead(rcvd_message_1, 4 + raw_suites_len + P256_ELEM_LEN);
-            if ead_res.is_ok() {
-                ead_1 = ead_res.unwrap();
-                error = EDHOCError::Success;
+                // check surplus bstr encoding
+                c_i = rcvd_message_1.content[3 + raw_suites_len + P256_ELEM_LEN];
+                let c_i_lookahead = rcvd_message_1.content[4 + raw_suites_len + P256_ELEM_LEN];
+                if !should_encoded_conn_id_be_minimal(c_i, c_i_lookahead) {
```

Comment:
We do, but as per Section [3.3.2.](https://www.ietf.org/archive/id/draft-ietf-lake-edhoc-22.html#section-3.3.2) there are two ways of representing the CID: byte-string or small-integer (-24 up to 23).
If a CID is a small-integer, it must be encoded as integer.
Here I check if the message contains a byte-string of size 1 whose single value is -24 up to 23.
We never generate that kind of CID, but other implementations could.

Reply:
I see. But your check is really tailored to the invalid test vector we have at hand. It should be general in the sense that we reject anything that is not supported by the application. We still do not support bstr-encoded CID if I am not mistaken, so you could only check whether the current byte is in the range -24 to 23. Anything else is a parsing error. What do you think?

Reply:
Yes, that makes sense. We indeed do not support bstr-encoded CID. Adjusted.

Reply:
I don't see how this was fixed, you are still using the "lookahead" value. Is that needed?

Reply:
It is still present in the Rust version, but I only applied the corrections in the hacspec version since that was where you commented first. I will change the hacspec one then modify the rust version.

Reply:
oops, sounds good!